subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java

```diff
@@ -194,8 +192,8 @@ FileSystemSnapshot snapshotDirectoryTree(final DirectoryFileTree dirTree) {
             public FileSystemSnapshot create() {
                 FileSystemLocationSnapshot snapshot = fileSystemMirror.getSnapshot(path);
                 if (snapshot == null) {
-                    snapshot = snapshotAndCache(dirTree.getDir(), patterns);
-                    return filterSnapshot(snapshot, EMPTY_PATTERN_SET);
+                    snapshot = snapshotAndCache(root, patterns);
+                    return snapshot.getType() != FileType.Directory ? filterSnapshot(snapshot, patterns) : filterSnapshot(snapshot, EMPTY_PATTERN_SET);
```

Comment:
Since the `DirectorySnapshotter` already takes the patterns into account - so the resulting file tree is already filtered. Only for files we need to filter the single file manually.